glib-macros/src/clone.rs

```diff
@@ -1,952 +1,487 @@
-                ")"
-            }
-        }
-        Delimiter::Brace => {
-            if open {
-                "{"
-            } else {
-                "}"
-            }
-        }
-        Delimiter::Bracket => {
-            if open {
-                "["
+                let next_attrs = &input.call(Attribute::parse_outer)?;
+                if !next_attrs.is_empty() {
+                    return Err(syn::Error::new_spanned(
+                        &attrs[0],
+                        "The `default_return` attribute must not be followed by more attributes",
+                    ));
+                }
+
+                break;
             } else {
-                "]"
+                return Err(syn::Error::new_spanned(
+                        attr,
+                        "Unsupported attribute: only `strong`, `weak`, `weak_allow_none`, `to_owned` and `default_return` are supported",
+                ));
             }
         }
-        Delimiter::None => "",
-    }
-}
 
-fn group_to_string(g: &Group) -> String {
-    format!(
-        "{}{}{}",
-        delimiter_to_string(g.delimiter(), true),
-        tokens_to_string(PeekableProcIter::from(g)),
-        delimiter_to_string(g.delimiter(), false),
-    )
-}
-
-#[doc(alias = "get_expr")]
-fn expr(parts: &mut PeekableProcIter) -> Result<String, TokenStream> {
-    let mut ret = String::new();
-    let mut total = 0;
-    let span = parts.current_span;
-    match parts.next() {
-        Some(TokenTree::Literal(l)) => ret.push_str(&l.to_string()),
-        Some(TokenTree::Ident(i)) => ret.push_str(&i.to_string()),
-        Some(TokenTree::Punct(p)) => match p.to_string().as_str() {
-            "[" | "{" | "(" => {
-                total += 1;
-            }
-            x => {
-                return Err(parts
-                    .generate_error(&format!("Unexpected token `{x}` after `@default-return`")))
-            }
-        },
-        Some(TokenTree::Group(g)) => return Ok(group_to_string(&g)),
-        None => {
-            return Err(
-                parts.generate_error_with_span("Unexpected end after `@default-return`", span)
-            )
-        }
-    };
-    loop {
-        match parts.peek() {
-            Some(TokenTree::Punct(p)) => {
-                let p_s = p.to_string();
-                if p_s == "{" || p_s == "(" || p_s == "[" || p_s == "<" {
-                    total += 1;
-                } else if p_s == "}" || p_s == ")" || p_s == "]" || p_s == ">" {
-                    total -= 1;
-                } else if p_s == "," && total == 0 {
-                    return Ok(ret);
+        // Following is a closure or async block
+        let expr = input.parse::<Expr>()?;
+        let body = match expr {
+            Expr::Async(async_) => {
+                if async_.capture.is_none() {
+                    return Err(syn::Error::new_spanned(
+                        async_,
+                        "Async blocks with the `move` keyword required",
```

Comment:
```suggestion
                        "Async blocks must be moved. Use the `move` keyword to do so",
```

Reply:
The block is not moved, the captured variables are

Reply:
Proof that the error message needs improvement. :wink:

Reply:
My message says exactly what is required: people have to add the `move` keyword.

Do you have a suggestion for a better message?